### PR TITLE
[CCFPCM-0374] round 3 POS heuristics (pt 1)

### PR DESCRIPTION
**Pull request by chelsea-EYDS**
[CCFPCM-0374](https://bcdevex.atlassian.net/browse/CCFPCM-0374)

Objective: 

- small refactor on reconcile lambda 
- pos reconciliation updated:
- separated heuristics into a different service
- added third heuristic
- updated tests

**TODO**

- we need to use or create an api for BC holidays in order to skip these days in our look back logic for pos reconciliation. Currently looking back one business day, but holidays will not be taken into account.

### CODE REVIEW

File: apps/backend/src/database/migrations/1683335003926-migration.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+
+export class migration1683335003926 implements MigrationInterface {
+  name = 'migration1683335003926';
+
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(
+      `ALTER TABLE "payment" ADD "heuristic_match_round" integer `
```
Comment (chelsea-EYDS) [inline apps/backend/src/database/migrations/1683335003926-migration.ts:8]: change to enum

File: apps/backend/src/deposits/entities/pos-deposit.entity.ts [hunk 2]
```diff
@@ -64,6 +64,9 @@ export class POSDepositEntity {
   @JoinColumn({ name: 'payment_method', referencedColumnName: 'method' })
   payment_method: Relation<PaymentMethodEntity>;
 
+  @Column({ nullable: true })
```
Comment (chelsea-EYDS) [inline apps/backend/src/deposits/entities/pos-deposit.entity.ts:67]: type enum

File: apps/backend/src/deposits/pos-deposit.service.ts [hunk 3]
```diff
@@ -22,31 +22,36 @@ export class PosDepositService {
...
-    date: string,
+    dateQuery: { yesterday: Date; today: Date },
     program: Ministries,
     location: LocationEntity,
-    status?: MatchStatus
+    status?: MatchStatus[]
   ): Promise<POSDepositEntity[]> {
-    const depositStatus = status ? status : In(MatchStatusAll);
+    const depositStatus = status ? status : MatchStatusAll;
+    const { yesterday, today } = dateQuery;
```
Comment (chelsea-EYDS) [inline apps/backend/src/deposits/pos-deposit.service.ts:31]: minDate maxDate

File: apps/backend/src/reconciliation/cash-reconciliation.service.ts [hunk 4]
```diff
@@ -116,6 +116,7 @@ export class CashReconciliationService {
           });
           break;
         }
+        break;
```
Comment (chelsea-EYDS) [inline apps/backend/src/reconciliation/cash-reconciliation.service.ts:119]: remove double break

File: apps/backend/src/reconciliation/pos-reconciliation.service.ts [hunk 5]
```diff
@@ -11,238 +12,299 @@ import { PaymentEntity } from '../transaction/entities/payment.entity';
...
+      status: MatchStatus.EXCEPTION
+    }));
 
+    return {
+      paymentExceptions,
+      depositExceptions
+    };
+  }
+
+  public logMatches(
```
Comment (chelsea-EYDS) [inline apps/backend/src/reconciliation/pos-reconciliation.service.ts:246]: revert to inline logging

File: apps/backend/src/reconciliation/pos-heuristics.ts [hunk 6]
```diff
@@ -0,0 +1,100 @@
+import { Injectable } from '@nestjs/common';
+import { differenceInBusinessDays, differenceInMinutes } from 'date-fns';
+import { MatchStatus } from '../common/const';
+import { POSDepositEntity } from '../deposits/entities/pos-deposit.entity';
+import { PaymentEntity } from '../transaction/entities/payment.entity';
+
+@Injectable()
+export class PosMatchHeuristics {
```
[the remark this answers is not in the document]
Comment (chelsea-EYDS) [reply apps/backend/src/reconciliation/pos-heuristics.ts:12]: I have added some more comments. Please let me know if they need more description, they might be too brief.

**Review by fw-noel — APPROVED** [review]

File: apps/backend/src/reconciliation/pos-heuristics.ts [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+ * @class
+ * @description This class contains the heuristics rounds for matching a payment to a deposit.
+ */
+@Injectable()
+export class PosMatchHeuristics {
+  /**
+   * Match payment to deposit on the same payment method
+   * @param {PaymentEntity} payment
+   * @param {POSDepositEntity} deposit
+   * @returns {boolen}
```
[the remark this answers is not in the document]
Comment (chelsea-EYDS) [reply apps/backend/src/reconciliation/pos-heuristics.ts:17]: lol, oops!

File: .prettierignore [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+.env.test.local
+.env.production.local
+
+# vercel
+.vercel
+
+#hbs
+*.hbs
+
+
```
Comment (chelsea-EYDS) [inline .prettierignore:48]: yml

File: apps/backend/src/reconciliation/pos-reconciliation.service.ts [hunk 9]
```diff
@@ -9,198 +14,144 @@ import { LocationEntity } from '../location/entities/master-location-data.entity
...
-      date,
+      dateRange,
       program,
       location,
-      MatchStatus.PENDING
+      [MatchStatus.PENDING, MatchStatus.IN_PROGRESS]
     );
 
-    if (pendingPayments.length === 0 && pendingDeposits.length === 0) {
+    if (pendingPayments.length === 0 || pendingDeposits.length === 0) {
```
Comment (chelsea-EYDS) [inline apps/backend/src/reconciliation/pos-reconciliation.service.ts:71]: double check "or"

**Review by fw-noel — DISMISSED** [review]

**Review by fw-noel — APPROVED** [review]

**Pull request merged**